include/ensmallen_bits/adamr/adamr.hpp

```diff
@@ -0,0 +1,170 @@
+ * @tparam DecayPolicy The StepSize decay policy to be used by default
+ * CyclicalDecay.
+*/
+template<typename UpdateRule = AdamUpdate,typename DecayPolicyType = CyclicalDecay>
+class AdamRType
+{
+ public:
+  /**
+   * Construct the AdamR optimizer with the given function and parameters. The
+   * defaults here are not necessarily good for the given problem, so it is
+   * suggested that the values used be tailored to the task at hand.  The
+   * maximum number of iterations refers to the maximum number of points that
+   * are processed (i.e., one iteration equals one point; one iteration does not
+   * equal one pass over the dataset).
+   *
+   * @param stepSize Maximum and initial step size for each batch of warm
+                     restart.
+   * @param epochRestart Restart Rate for Warm Restarts
+   * @param multFactor Multiplier for epochRestart
+   * @param stepSizeMin Minimum and final step size for each batch of warm
+                        restart (Use DecayPolicy() method with CyclicalDecay).
+   * @param batchSize Number of points to process in a single step.
+   * @param beta1 Exponential decay rate for the first moment estimates.
+   * @param beta2 Exponential decay rate for the weighted infinity norm
+                  estimates.
+   * @param eps Value used to initialise the mean squared gradient parameter.
+   * @param maxIterations Maximum number of iterations allowed (0 means no
+   *                      limit).
+   * @param tolerance Maximum absolute tolerance to terminate algorithm.
+   * @param shuffle If true, the function order is shuffled; otherwise, each
+   *        function is visited in linear order.
+   * @param resetPolicy If true, parameters are reset before every Optimize
+   *        call; otherwise, their values are retained.
+   * @param updateRule Update Policy to be used for AdamR.
+   * @param decayPolicy Decay Policy to be used for AdamR.
+   */
+  AdamRType(const double stepSize = 0.001,
+            const size_t epochRestart = 50,
+            const double multFactor = 2.0,
+            const size_t batchSize = 32,
+            const double beta1 = 0.9,
+            const double beta2 = 0.999,
+            const double eps = 1e-8,
+            const size_t maxIterations = 100000,
+            const double tolerance = 1e-5,
+            const bool shuffle = true,
+            const bool resetPolicy = true);
+
+  /**
+   * Optimize the given function using AdamR. The given starting point will be
+   * modified to store the finishing point of the algorithm, and the final
+   * objective value is returned.
+   *
+   * @tparam DecomposableFunctionType Type of the function to optimize.
+   * @param function Function to optimize.
+   * @param iterate Starting point (will be modified).
+   * @return Objective value of the final point.
+   */
+  template<typename DecomposableFunctionType>
+  double Optimize(DecomposableFunctionType& function, arma::mat& iterate);
```

Comment:
There should be an empty line below this `Optimize()` declaration and the next method.